### ISSUE TITLE
Add documentation for GITHUB_TOKEN requirement

### ISSUE THREAD

**Issue by malaterre** [cpp-linter/cpp-linter-action#136]
The documentation does not make it very clear what type of access is required:

1. fine-grained personal access token ?
2. personal access token (classic) ?

What type of permissions ?

Thanks

**Comment by 2bndy5** [comment]
This mainly because it was unclear from the REST API docs what exact permission it needed to execute.

You are correct though. We should mention explicitly what permissions are used when the token is required.

**Comment by 2bndy5** [comment]
Just so you can see what vagueness I'm working with:
1. REST API list of endpoints for [commits](https://docs.github.com/en/rest/overview/endpoints-available-for-fine-grained-personal-access-tokens?apiVersion=2022-11-28#commits) and [issues](https://docs.github.com/en/rest/overview/endpoints-available-for-fine-grained-personal-access-tokens?apiVersion=2022-11-28#issues) which is used for PR threads also. The [pulls](https://docs.github.com/en/rest/overview/endpoints-available-for-fine-grained-personal-access-tokens?apiVersion=2022-11-28#pulls) endpoints are about PR reviews which are much more complex to implement, so we don't use the pulls endpoints.
2. We use the following REST API when tokens are needed:
   - For a PR:
      - [`GET /repos/{owner}/{repo}/issues/{issue_number}/comments`](https://docs.github.com/en/rest/issues#list-issue-comments) (needed for fetching existing comments)
      - [`DELETE /repos/{owner}/{repo}/issues/comments/{comment_id}`](https://docs.github.com/en/rest/issues#delete-an-issue-comment) (used to delete comments that this action creates)
      - [`POST /repos/{owner}/{repo}/issues/{issue_number}/comments`](https://docs.github.com/en/rest/issues#create-an-issue-comment) (used to create a new thread comment)
   - For a commit (push event):
      - [`GET /repos/{owner}/{repo}/commits/{commit_sha}/comments`](https://docs.github.com/en/rest/commits/comments#list-commit-comments) (needed for fetching existing comments)
      - [`DELETE /repos/{owner}/{repo}/comments/{comment_id}`](https://docs.github.com/en/rest/commits/comments#delete-a-commit-comment) (used to delete comments that this action creates)
      - [`POST /repos/{owner}/{repo}/commits/{commit_sha}/comments`](https://docs.github.com/en/rest/commits/comments#create-a-commit-comment) (used to create a new thread comment)

**Comment by 2bndy5** [comment]
> 1. fine-grained personal access token ?

This feature is still in beta and lacks adequate documentation to answer the question. While it might work with this action, we won't be held responsible if it doesn't.

---
From what I can gather, I think the following permissions are used when a token is required (for a public repo). Note, this is mostly guesses since the REST API doesn't always mention for what the token needs permission.

```yml
permissions:
  checks: read|write
  # write for our file-annotations option (I think - probably not)

  pull-requests: read|write 
  # write to remove duplicate action comments and post new comments
  # read to get a list of changed files and a list of existing comments

  repository-projects: read|write  # (for push events)
  # write to remove duplicate action comments and post new comments
  # read to get list of changed files and a list of existing comments
```
See also [Assigning permissions to jobs](https://docs.github.com/en/actions/using-jobs/assigning-permissions-to-jobs).

Private repo access seems largely undocumented, but we have [reportedly] run into permission problems when getting a list of PR comments and a list of changed files.

---
I think the only way we can be certain what permissions are needed is by trial and error (we have a test repo for that). But my time is currently limited, and I would very much appreciate user feedback on what permission settings worked and what didn't.

**Comment by 2bndy5** [comment]
>Could also comment, why at least one alternative solution works out of the box (no need for GITHUB_TOKEN setup):
>
> -   https://github.com/marketplace/actions/clang-tidy-review#inputs

They use the token with implied consent, and it is unclear what they need it for without delving into the source code.

We refuse to confiscate permission without user consent. That's why we won't use a default value for the token. It is better to know when your permission is needed for something, instead of giving us the house keys and trusting we don't burn the house down.

**Comment by 2bndy5** [comment]
It seems the GIthub REST API docs have been updated again. They now explicitly state what permissions are needed for each REST API endpoint.

Re-opened this as it should be addressed in #198
